### PR TITLE
Fly us directly into the sun (BYOND init order)

### PR DESCRIPTION
**Pull request by Cyberboss**
I tried, I really did.

At the drop of a hat BYOND can completely switch up the order static initializers are run in.
We can't win.

I did the next best thing.

Every static initializer is wrapped in a macro `STATIC_INIT(X)` (there's a new grep to enforce this) that runs the expressiong `world.Genesis() ? null : (##X)`. There's now a guard in `world.Genesis()` that prevents it from being called incorrectly multiple times.

image

### CODE REVIEW

**Comment by optimumtact** [comment]
complex overly fragile change for the purposes of a dev tool that three people use

bold

**Comment by ZephyrTFA** [comment]
Ain't no way in hell I'm looking through this entire thing

File: code/game/objects/items/pillow.dm [hunk 1]
```diff
@@ -10,7 +10,7 @@
 	w_class = WEIGHT_CLASS_NORMAL
 	damtype = STAMINA
 	///change the description based on the pillow tag
-	var/static/tag_desc = "This one seems to have its tag removed."
+	var/static/tag_desc = STATIC_INIT("This one seems to have its tag removed.")
```
Comment (LemonInTheDark) [inline code/game/objects/items/pillow.dm:13]: redundant, right?

File: code/modules/admin/verbs/SDQL2/SDQL_2.dm [hunk 2]
```diff
@@ -979,7 +979,7 @@ GLOBAL_DATUM_INIT(sdql2_vv_statobj, /obj/effect/statclick/sdql2_vv_all, new(null
 	return querys
 
 /proc/SDQL_testout(list/query_tree, indent = 0)
-	var/static/whitespace = "&nbsp;&nbsp;&nbsp; "
+	var/static/whitespace = STATIC_INIT("&nbsp;&nbsp;&nbsp; ")
```
Comment (LemonInTheDark) [inline code/modules/admin/verbs/SDQL2/SDQL_2.dm:982]: similarly redundant? unless it's somehow worth while to hang a ref to a string permanently?

File: code/__DEFINES/_globals.dm [hunk 3]
```diff
@@ -60,3 +60,6 @@
 
 /// Create a typed null global
 #define GLOBAL_DATUM(X, Typepath) GLOBAL_RAW(Typepath/##X); GLOBAL_UNMANAGED(X)
+
+/// Used to determine BYOND static init order
+#define STATIC_INIT(expression) world._(##expression)
```
Comment (LemonInTheDark) [inline code/__DEFINES/_globals.dm:65]: do we need this if profiling tools aren't in use?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
cursed as hell

File: code/__DEFINES/_globals.dm (hunk 3 above)
Comment (Cyberboss) [reply code/__DEFINES/_globals.dm:65]: Depends on how much control we want over init order.

With it as it is we can (%99.99) guarantee Master and GLOB will be initialized before statics run.

**Comment by Cyberboss** [comment]
> Ain't no way in hell I'm looking through this entire thing

Ignore the three pushes where I go commit regex

File: code/modules/admin/verbs/SDQL2/SDQL_2.dm (hunk 2 above)
Comment (Cyberboss) [reply code/modules/admin/verbs/SDQL2/SDQL_2.dm:982]: Yes, but out of scope

**Comment by optimumtact** [comment]
> > complex overly fragile change for the purposes of a dev tool that three people use
> > bold
> 
> I didn't really do this for tracy. I did it for more determinism during init.

how does it provide determinism during init, if the statics can be in any order

**Comment by Cyberboss** [comment]
In the form of knowing that we can decide what the first code to run is and not some random static

**Comment by Mothblocks** [comment]
:-1: yeah i'm 100% not on board with this at all. one of the highest priorities for dev cycles initiative is to be as invisible as possible, other than for faster cycles. the times when we do it (ex: `RegisterSignals`) are only done if the gains are significant enough. if this is just for making profiling easier, then i'm against it.

- statics make up less than 100ms of total startup cost
- once we figured out a trick, it took us ~10 minutes to figure out the exact costs of everything. if we made a hackmd page about how to do it it would take just a few minutes
- we could most likely manually do that whenever we think about it and get equivalent results

@LemonInTheDark what do you think? i can't see myself doing anything other than rejecting this

**Comment by Mothblocks** [comment]
not to mention the most expensive work is already going to be visible in the byond profiler, which we do check sometimes.

i would much rather we put our energy into a byond feature request for static initializing being done in order of definition

**Comment by LemonInTheDark** [comment]
This is nice for profiling, but you're right it's too intrusive if that's its only purpose
Dominion seems to be also targeting a concrete init order, particularly ensuring globs and such are created first thing.
This would be useful, and would prevent some stupid gottchas in the code (like glob sometimes just not existing in very early execution)

I'm not sure if that's worth the macro around each static use. I do think we should bully lummox about this, just generally tho.

**Comment by Mothblocks** [comment]
Honestly if we were interested in tackling any problems here (which I am not without more concrete evidence that it's an issue), having an interface that's like C statics (run once, but when the function is called) would be far more useful. Something like:

```dm
/proc/do_thing()
    var/expensive_to_compute = cache_result(PROC_REF(calculate_expensive_thing))

    // keep working
```

**Comment by Mothblocks** [comment]
But yeah I think this development hit is not worth the far-too-theoretical gains

**Comment by Cyberboss** [comment]
Force updating this just because I found a better way that doesn't have the drawback of missing one expression

**Pull request closed without merging**